util/azure/utils.go

```diff
@@ -0,0 +1,352 @@
+	ActionInfo           AuthorizationActionInfo
+	IsNamespacedResource bool
+}
+
+type VerbAndActionsMap map[string]DataAction
+
+func NewVerbAndActionsMap() VerbAndActionsMap {
+	return make(map[string]DataAction)
+}
+
+type ResourceAndVerbMap map[string]VerbAndActionsMap
+
+func NewResourceAndVerbMap() ResourceAndVerbMap {
+	return make(map[string]VerbAndActionsMap)
+}
+
+type OperationsMap map[string]ResourceAndVerbMap
+
+func NewOperationsMap() OperationsMap {
+	return make(map[string]ResourceAndVerbMap)
+}
+
+func (o OperationsMap) String() string {
+	opMapString, _ := json.Marshal(o)
+	return string(opMapString)
+}
+
+/*
+   FetchListOfResources does the following:
+   1. Fetches list of ApiResources from the apiserver
+   2. Fetches list of Data Actions via Get Operations call on Azure
+   3. creates OperationsMap which is a map of "group": { "resource": { "verb": DataAction{} } } }
+   This map is used to create list of AuthorizationActionInfos when we get a SAR request where Resource/Verb/Group is *
+*/
+func FetchListOfResources(clusterType string, environment string, loginURL string, kubeconfigFilePath string, tenantId string, clientID string, clientSecret string) (OperationsMap, error) {
+	operationsMap := OperationsMap{}
+	apiResourcesList, err := fetchApiResources(kubeconfigFilePath)
+	if err != nil {
+		return operationsMap, errors.Wrap(err, "Failed to fetch list of api-resources from apiserver.")
+	}
+
+	operationsList, err := fetchDataActionsList(environment, clusterType, loginURL, tenantId, clientID, clientSecret)
+	if err != nil {
+		return operationsMap, errors.Wrap(err, "Failed to fetch operations from Azure.")
+	}
+
+	operationsMap = createOperationsMap(apiResourcesList, operationsList, clusterType)
+
+	klog.V(5).Infof("Operations Map created for resources: %s", operationsMap)
+
+	return operationsMap, nil
+}
+
+func createOperationsMap(apiResourcesList []*metav1.APIResourceList, operationsList []Operation, clusterType string) OperationsMap {
+	operationsMap := NewOperationsMap()
+
+	for _, resList := range apiResourcesList {
+		if len(resList.APIResources) == 0 {
+			continue
+		}
+
+		group := "v1" // core api group
+		if resList.GroupVersion != "" && resList.GroupVersion != "v1" {
+			group = strings.Split(resList.GroupVersion, "/")[0]
+		}
+
+		for _, apiResource := range resList.APIResources {
+			if strings.Contains(apiResource.Name, "/") {
+				continue
+			}
+
+			actionId := clusterType
+			if group != "v1" {
+				actionId = path.Join(actionId, group)
+			}
+
+			resourceName := apiResource.Name
+
+			actionId = path.Join(actionId, resourceName)
+
+			for _, operation := range operationsList {
+				if strings.Contains(operation.Name, actionId) {
+					opNameArr := strings.Split(operation.Name, "/")
+
+					if group != "v1" {
+						// extra validation to make sure groups are the same
+						if group != opNameArr[2] {
+							continue
+						}
+					} else {
+						// make sure resources are the same for core apigroup
```

Comment:
Pls add comments why have we added this validation with example.

Reply:
done